### PR TITLE
[FEATURE] Add createEventHub

### PR DESCRIPTION
**Pull request by atomiks**
## Description
A tiny pubsub implementation

## What does your PR belong to?
- [ ] Website
- [x] Snippets
- [ ] General / Things regarding the repository (like CI Integration)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking improvement of a snippet)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have checked that the changes are working properly
- [x] I have checked that there isn't any PR doing the same
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: snippets/createEventHub.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+### createEventHub
+
+Creates a pubsub event hub with `emit`, `on`, and `off` methods.
```
Comment (Chalarangelo) [inline snippets/createEventHub.md]: `pubsub` might not be apparent to some readers, is there a better term?

File: tag_database [hunk 2]
```diff
@@ -20,6 +20,7 @@ compose:function
 copyToClipboard:browser
 countOccurrences:array
 countVowels:string
+createEventHub:utility
```
Comment (Chalarangelo) [inline tag_database]: Tags should be `browser,utility,advanced`.

File: tag_database [hunk 3]
```diff
@@ -20,7 +20,7 @@ compose:function
 copyToClipboard:browser
 countOccurrences:array
 countVowels:string
-createEventHub:utility
+createEventHub:browser,utility,advanced
```
Comment (Chalarangelo) [inline tag_database]: 👍

**Comment by atomiks** [comment]
@skatcat31 in that case we can replace the plain object with a Map?

**Comment by skatcat31** [comment]
@atomiks I feel like it went over your head possibly. What I meant is
```js
hub.emit('message', { hello: 'world'}); // this is valid in this implementation, would require type checking and throwing otherwise
```

**Comment by atomiks** [comment]
😅 yeah I was confused

But, wait.. why should an object as the data throw an error? That allows you to access multiple data points from a single argument

**Comment by skatcat31** [comment]
@atomiks that's not what I meant either. I just meant that you should note in the description/example that the event can be of any type subscribed to the topic.

As for the topic being an object I'm not certain I like that functionality from Map as it goes against the idea of being able to provide new instances of a string to trigger the event. The problem becomes if someone does
```js
emit(String('test'), 'test1');
emit(String('test'), 'test2');
```

**Comment by skatcat31** [comment]
Right so some more clarification:

What I'm saying is that you should show the example
```js
hub.emit('message', {hello:'world'})
```
and note that the event is anything in the description. Sorry my last comment was a bit esoteric

**Comment by atomiks** [comment]
I think events should only be strings, doesn't really make sense to be anything else? I made the example use an object now... not 100% sure what you mean tho

File: snippets/createEventHub.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  off(event, handler) {
+    const i = (this.hub[event] || []).findIndex(h => h === handler);
+    if (i > -1) this.hub[event].splice(i, 1);
+  }
+});
+```
+
+```js
+const fn = data => console.log(data.hello);
+const hub = createEventHub();
```
Comment (skatcat31) [inline snippets/createEventHub.md]: I'm guessing I'm not explaining it very well but what I was trying to get at is that the example should be
```js
const fn = data => console.log(data);
const hub = createEventHub();

hub.on('message', fn); // subscribe the handler to listen for 'message' events
hub.emit('message', { hello: 'world' }); // logs the object
hub.emit('message', 'hello world' }); // logs 'hello world'
hub.off('message', fn); // unsubscribe the handler from 'message' events
```

Comment (skatcat31) [reply snippets/createEventHub.md]: I've seen so many developer not understand that the type to the event handler is whatever is passed to the event. I'm not sure why I've seen it so many times, but I have enough that it should be here

**Comment by skatcat31** [comment]
@Chalarangelo all of my concerns are met

**Comment by Chalarangelo** [comment]
I'll review this again in the morning (mainly cleanup and a final test just in case) and merge. Seems ok to me!

**Pull request merged**